### PR TITLE
[FIX] stock: UoM of a product should not be changed when done stock

### PR DESCRIPTION
**Pull request by wtaferner**
moves exist. 

Backport from saas-6 (242a76cc979c3fc237d9e3c44ad18361b33a0d98)
## Impacted versions:

8.0
## Steps to reproduce:
- Create some stock moves with any product with a dedicated UoM Unit
- Change the UoM to Dozens
- Create more stock moves with this product
## Current behavior:
- Quants do not have a dedicated UoM and therefore always rely on the fact that the product UoM does represent its quantity which is leading to inconsistencies in the use case described above. No transformation is done on user configuration change and input
## Expected behavior:
- The user should be disallowed to change the UoM as soon it has stock moves even if it is from the same category.

Resolves OPW-643168

### CODE REVIEW

**Comment by beledouxdenis** [comment]
Hello Wolfgang, 

I do not know yet how it works in saas-6, but in Odoo 8.0 you are allowed to change the unit of measure of a product if, and only if, the new uom is part of the same category than the old uom. See below screenshot, illustrating the fact such kind of change is already prevented, without your PR.
odoo


This check is done here:
https://github.com/odoo/odoo/blob/8.0/addons/product/product.py#L756

To me, you should always be allowed to change the unit of measure of a product as long as the unit of measure category is the same (e.g. from `Unit(s)` to `Dozen(s)`, but not from `Unit(s)` to `cm`). Because, when you change of uom within a same category, you are still able to compute the current stock levels (you can convert the stock moves done in `Unit(s)` to `Dozen(s)`), but you cannot compute the inventory when you change of category (`Unit('s)` to `cm`).

The fact such kind of change within a same category is prevented in saas-6 might in fact be the real issue.

Regards,

**Comment by beledouxdenis** [comment]
Isn't the issue the fact the quants do not have their own uom field ? So we can compute the inventory according to the quants quantity/uom ?

**Comment by wtaferner** [comment]
Hi Denis,
Believe me it is already an issue in v8 as I had a real life example and it was embarrassing ;-)

But I try to explain:
Quants are taken for calculation of the current and future inventory and they live as long they are not neutralized without a dedicated UoM as they rely on the product configuration (FIFO, etc.)

If the DAU do change the UoM of a product the correspondent quants are not changed and the next time I do a stock move the UoM assumes that the quants were ALWAYS be calculated on the UoM in the given configuration. Based on this you could get nasty inconsistency and therefore wrong data in your instance.

You can check back with @jco-odoo, but I am almost 100% sure that this fix is for the good and not avoidable ;-)

Let me know if I can help further.

Best regards,
Wolfgang

**Comment by beledouxdenis** [comment]
Well then, let's wait for @jco-odoo confirmation that this has to be back-ported in Odoo 8.0.

Nevertheless, I still believe this should be solved in master by adding a many2one field to `product.uom` on the `stock.quant` model, and performing the unit of measure ratio computation when the units of measure aren't the same for all `stock.quant` for a same `product.product`.

I guess it could even be done in stable with a computed field guessing the `stock.quant` from its `stock.move` history.

**Comment by wtaferner** [comment]
Hmm, this would be even better to get it completely right, but I guess Jos had his reasons why it was not introduced like this, but I agree that this would be the ultimate solution ;-) Anyway, a fix must be done for version 8 too :-)

**Comment by jco-odoo** [comment]
Quants are always in the default UoM.  That way they are also rounded consistently and that is one of the reasons that we did not implement a conversion to another UoM.  If you allow two UoMs, they will be rounded differently and you need to invent a bunch of rules to determine when they will be rounded in which way.

File: addons/stock/product.py [hunk 1]
```diff
@@ -482,6 +483,16 @@ def action_view_stock_moves(self, cr, uid, ids, context=None):
...
         return result
 
+    def write(self, cr, uid, ids, vals, context=None):
+        if 'uom_id' in vals:
+            new_uom = self.pool.get('product.uom').browse(cr, uid, vals['uom_id'], context=context)
+            for product in self.browse(cr, uid, ids, context=context):
+                old_uom = product.uom_id
+                if old_uom != new_uom:
+                    if self.pool.get('stock.move').search(cr, uid, [('product_id', 'in', [x.id for x in product.product_variant_ids]), ('state', '=', 'done')], limit=1, context=context):
+                        raise except_orm(_('Warning'), _("You can not change the unit of measure of a product that has already been used in a done stock move. If you need to change the unit of measure, you may deactivate this product."))
```
Comment (jco-odoo) [inline addons/stock/product.py]: except_orm to UserError?

Comment (wtaferner) [inline addons/stock/product.py]: Does not exist in v8 yet. At least I did not find it :-)

**Comment by beledouxdenis** [comment]
To me, that would lead to a huge regression, the inability to change the unit of measure of a same unit of measure category. This was possible in 7.0 and the ratio computation were done correctly.

Unless you have a specific example, the rounding issue you mention doesn't stand.

**Comment by beledouxdenis** [comment]
We just had a talk: This has nothing to do with a rounding issue, but the fact the `stock.move` model stores a field containing the quantity within the product default uom:
https://github.com/odoo/odoo/blob/8.0/addons/stock/stock.py#L1742-L1743

Changing the uom of a product, even in the same category, would therefore lead to the re-computation of this field for all existing moves for that product, which can be very problematic for huge databases.

I therefore accept this PR. I will merge it right away.

**Comment by wtaferner** [comment]
@beledouxdenis 
Thank you for the follow-up and I appreciate your efforts and watching the regression panel ;-)

**Pull request closed without merging**